File: 1. Create and print a singly linked list.cpp
/*
 Task: Create and print a singly linked list
 Time Complexity: append O(n) (as implemented)
 Space Complexity: O(n)
*/

#include <iostream>

using namespace std;

struct Node
{
    int data;
    Node* next;

    Node(int d)
        : data(d)
        , next(nullptr)
    {
    }
};

void append(Node*& head, int val)
{
    Node* n = new Node(val);
    if (!head)
    {
        head = n;
        return;
    }

    Node* cur = head;
    while (cur->next)
    {
        cur = cur->next;
    }

    cur->next = n;
}

void printList(Node* head)
{
    while (head)
    {
        cout << head->data;
        if (head->next)
        {
            cout << " -> ";
        }
        head = head->next;
    }

    cout << endl;
}

int main()
{
    Node* head = nullptr;
    append(head, 10);
    append(head, 20);
    append(head, 30);

    cout << "Singly linked list: ";
    printList(head);
    return 0;
}

// sc tc

File: 1. Create and print a singly linked list.py
"""
Task: Create and print a singly linked list
Time Complexity: depends on operation (append O(n) as implemented)
Space Complexity: O(n) for the list
"""

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def append(self, val):
        node = Node(val)
        if not self.head:
            self.head = node
            return
        cur = self.head
        while cur.next:
            cur = cur.next
        cur.next = node

    def print_list(self):
        cur = self.head
        out = []
        while cur:
            out.append(str(cur.data))
            cur = cur.next
        print(' -> '.join(out))

if __name__ == '__main__':
    ll = LinkedList()
    ll.append(10)
    ll.append(20)
    ll.append(30)
    print('Singly linked list:', end=' ')
    ll.print_list()

# Task: Create and print a singly linked list
